### DISCUSSION TITLE
Avoid shipping raw `.ts` to node_modules

### DISCUSSION THREAD

**Discussion by ssalbdivad** [withastro/starlight#1891] [Feature Requests]
### What version of `starlight` are you using?

0.23.0

### What is your idea?

Currently, Starlight publishes raw `.ts` source to npm.

This would replace those with class `.d.ts` and `.js`.

### Why is this feature necessary?

Currently, I cannot edit my docs without getting a large number of extraneous errors that get pulled into my editor:

![image](https://github.com/withastro/starlight/assets/10645823/1d535c52-bcac-4a1e-b9f6-94f86657a963)

That behavior may be incorrect in terms of the TS LSP + editor integration since it does not occur when running `tsc` directly as `node_modules` is excluded, but this is just one of many problems that shipping raw `.ts` source causes.

It is impossible to guarantee your source will be compatible with user tsconfigs, and these errors can manifest in all sorts of obnoxious ways as I've encountered myself dealing with many reports of problems when I tried shipping `.ts` alongside my `.d.ts` in ArkType.

Hopefully this will be viable soon, but for now, due to fragility and performance issues, shipping `.ts` source should be avoided.

In a Discord discussion, @delucis mentioned there were some barriers around integrating `.d.ts` with `.astro` files. It would be a huge ROI for me and I'd expect many others to not be constantly contending with related issues, and very quickly for maintainers as well to avoid a deluge of tsconfig-related issues your growing userbase will encounter.

### Do you have examples of this feature in other projects?

Most projects do not ship raw `.ts`.

I've created a well-optimized build + dev config in [ArkType's monorepo](https://github.com/arktypeio/arktype) if that is helpful to refer to, though I suspect most of the challenge will be in the `.astro` integration which I know nothing about. Hoping it's not too thorny 🙏 

### Participation

- [ ] I am willing to submit a pull request for this feature.

**Comment by HiDeoo** [comment]
Adding a link to https://github.com/withastro/starlight/discussions/983 for reference, but I think it's fine to keep this discussion opened too as it adds a lot of information, e.g. I was personally not aware of this "Problems" tab in VSCode.

**Comment by Fryuni** [comment]
Shipping JS and d.ts should not be too difficult. We would still ship the original TS files in the package since that is needed for stack traces to provide the original position and allow debuggers to step through the original code instead of the transpiled code.

@delucis what barriers would those be? The packages I have that include both Astro components and TS files have been working fine and are distributed transpiled. Some Starlight-specific blocker?

**Reply by delucis** [reply]
I think mainly that `tsc` maybe can’t compile the [components façade module](https://github.com/withastro/starlight/blob/main/packages/starlight/components.ts) that re-exports Astro components?

Also not 100% sure how things that our Astro components _import_ will behave — that would also point towards maybe needing to keep the `.ts` files around? Not sure.

**Reply by ssalbdivad** [reply]
@Fryuni Shipping the TS files along the others would still be a huge improvement, but I actually tried doing the same thing in [ArkType](https://github.com/arktypeio/arktype) and got lots of complaints from users with incompatible tsconfig settings about errors they were seeing in `node_modules`.

On the other hand, I didn't hear a lot of feedback about it being very helpful to be able to debug the library's TS source.

I'm sure there are other factors that could tip the scales, but just wanted to share the results I had trying the same approach.

**Comment by stereobooster** [comment]
Shouldn't it be solved by ignoring node modules? Something like this:

```
{
  "compilerOptions": {
    "skipLibCheck": true
  },
}
```

**Reply by HiDeoo** [reply]
This [discussion](<https://github.com/withastro/starlight/discussions/983>), and the linked TS issue inside it explains in details why this approach would not work as [`skipLibCheck`](https://www.typescriptlang.org/tsconfig/#skipLibCheck) is only intended for typing files and not for `.ts` files:

> Skip type checking of declaration files.

**Reply by ssalbdivad** [reply]
Yes, unfortunately these settings and their defaults are generally very unintuitive.

In general, most people would expect `.ts` and `.d.ts` in their own code to be checked, while any files in `node_modules` would not be, but that is not only not the default, it's not an option at all. Even specifically adding `node_modules` to `exclude` often results in `.ts` files getting pulled in if they are somehow referenced indirectly 😔

**Comment by HiDeoo** [comment]
[Starlight 0.42](https://github.com/withastro/starlight/releases/tag/%40astrojs%2Fstarlight%400.42.0) now ships JavaScript files with dedicated type declaration files instead of TypeScript source files.
